packages/sqlite/src/index.ts

```diff
@@ -118,115 +92,91 @@ class DatabaseConnectionImplementation implements DatabaseConnection {
+      // however I do not know how to get
+      this._begin.run();
       try {
-        const result = fn(
+        const result = await fn(
           new DatabaseTransactionImplementation(this._database),
         );
-        await new Promise<void>((resolve, reject) => {
-          this._database.run('COMMIT', (err) => {
-            if (err) reject(err);
-            else resolve();
-          });
-        });
+        this._commit.run();
         return result;
       } catch (ex) {
-        await new Promise<void>((resolve, reject) => {
-          this._database.run('ROLLBACK', (err) => {
-            if (err) reject(err);
-            else resolve();
-          });
-        });
+        this._rollback.run();
         throw ex;
       }
     });
   }
 
   async dispose() {
-    await new Promise<void>((resolve, reject) => {
-      this._database.close((err) => {
-        if (err) reject(err);
-        else resolve();
-      });
-    });
+    this._database.close();
   }
 }
 
 export default function connect(
   filename: string = IN_MEMORY,
-  options: DatabaseConnectionOptions = {},
+  options: DatabaseOptions = {},
 ): DatabaseConnection {
   return new DatabaseConnectionImplementation(filename, options);
 }
 module.exports = Object.assign(connect, {
   default: connect,
-  DatabaseConnectionMode,
   IN_MEMORY,
   sql,
   isSqlQuery,
 });
 
 async function runQuery(
   query: SQLQuery,
-  database: sqlite.Database,
+  database: Database,
   lock: <T>(fn: () => Promise<T>) => Promise<T>,
 ) {
   const {text, values} = query.format(sqliteFormat);
   return lock(async () => {
-    return await new Promise<any[]>((resolve, reject) => {
-      database.all(text, values, (err, rows) => {
-        if (err) reject(err);
-        else resolve(rows);
-      });
-    });
+    const stm = database.prepare(text);
+    try {
+      return stm.all(...values);
+    } catch (_err) {
+      const err = _err as Error;
+      // TODO we should be able to catch this before calling all()
+      if (err.message.indexOf('This statement does not return data') >= 0) {
+        return stm.run(...values);
+      }
+      throw err
+    }
   });
 }
 
-interface Queue<T> {
-  push(item: T): void;
-  pop(): Promise<T>;
-  /**
-   * Amount of items in the queue
-   * This can be negative if pop has been called more times than push.
-   */
-  length: number;
-}
 async function* runQueryStream(
   query: SQLQuery,
-  database: sqlite.Database,
+  database: Database,
   lock: <T>(fn: () => Promise<T>) => Promise<T>,
 ): AsyncIterableIterator<any> {
-  const queue: Queue<{done: false; value: any} | {done: true; err: any}> =
-    new Queue();
   const {text, values} = query.format(sqliteFormat);
+  // TODO figure out how to avoid the extra closures
+  let resolve = () => {};
+  let reject: (reason? : any) => void;
+  reject = () => {};
+  const promise = new Promise<void>((res, rej) => {
+    resolve = res
+    reject = rej
+  })
```

Comment:
Given that you'd only ever do a streaming query for a very expensive query, I think the extra few closures shouldn't really matter at all.

Reply:
I had to create all those closure just to satisfy typescript. If you can suggest a way to let it be happy without instantiating `resolve` and `reject`, I would be _really happy_.